### PR TITLE
[Merged by Bors] - feat: induction principle for `ENNReal`-valued measurable functions from a sigma-finite space

### PR DESCRIPTION
**Pull request by YaelDillies**
From GibbsMeasure

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by RemyDegenne** [comment]
A function that can be approached by finite measure simple function is [MeasureTheory.FinStronglyMeasurable](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.html#MeasureTheory.FinStronglyMeasurable)

The result [MeasureTheory.finStronglyMeasurable_iff_measurable](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.html#MeasureTheory.finStronglyMeasurable_iff_measurable) states that it is equivalent to measurability in a space with second countable topology and a sigma-finite measure.

Could those results be used in your applications?

**Comment by YaelDillies** [comment]
Yes and no:
* I really need the explicit sequence of finite mass functions as I use it to define the Lesbegue conditional [here](https://github.com/james18lpc/GibbsMeasure/blob/main/GibbsMeasure/Prereqs/LebesgueCondExp.lean#L12-L23).
* I am happy to define it for `FinStronglyMeasurable` functions instead, if you ask.
* I don't need the induction principle `Measurable.ennreal_sigmaFinite_induction`. I will remove it if you ask.

**Comment by RemyDegenne** [comment]
What makes the sequence you get from FinStronglyMeasurable unfit for your purpose? Does it lack some property?([MeasureTheory.FinStronglyMeasurable.approx](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.html#MeasureTheory.FinStronglyMeasurable.approx))

**Comment by YaelDillies** [comment]
I just didn't know it existed! That solves my first two bullet points. What about the third?

**Comment by RemyDegenne** [comment]
FinStronglyMeasurable is not used much, and well hidden inside the StronglyMeasurable file, so it's no surprise you did not know about it!

For the third point, I'd say that the induction principle could be useful and I would be happy to merge it.

**Comment by YaelDillies** [comment]
I thought I could generalise the induction principle to any `FinStronglyMeasurable` function but actually I'm not sure. Here is the statement I attempted (to be added on L.1881 of `MeasureTheory.Function.StronglyMeasurable.Basic`):
````
variable {mα : MeasurableSpace α} {μ : Measure α}

/-- To prove something for an arbitrary `μ`-finite strongly measurable function into `ℝ≥0∞`, it
suffices to show that the property holds for (multiples of) characteristic functions with finite
`μ`-mass and is closed under addition and supremum of increasing sequences of functions.

It is possible to make the hypotheses in the induction steps a bit stronger, and such conditions
can be added once we need them (for example in `h_add` it is only necessary to consider the sum of
a simple function with a multiple of a characteristic function and that the intersection
of their images is a subset of `{0}`. -/
@[elab_as_elim]
lemma FinStronglyMeasurable.ennreal_induction {P : (α → ℝ≥0∞) → Prop}
    (h_ind : ∀ (c : ℝ≥0∞) ⦃s⦄, MeasurableSet s → μ s < ∞ → P (Set.indicator s fun _ => c))
    (h_add :
      ∀ ⦃f g : α → ℝ≥0∞⦄,
        Disjoint (support f) (support g) → Measurable f → Measurable g → P f → P g → P (f + g))
    (h_iSup :
      ∀ ⦃f : ℕ → α → ℝ≥0∞⦄, (∀ n, Measurable (f n)) → Monotone f → (∀ n, P (f n)) →
        P fun x => ⨆ n, f n x)
    ⦃f : α → ℝ≥0∞⦄ (hf : FinStronglyMeasurable f μ) : P f := by
  sorry
```

File: Mathlib/MeasureTheory/Function/SimpleFunc.lean [hunk 1]
```diff
@@ -1194,9 +1215,35 @@ theorem Measurable.ennreal_induction {α} [MeasurableSpace α] {P : (α → ℝ
...
+some sigma-finite measure and is closed under addition and supremum of increasing sequences of
+functions.
+
+It is possible to make the hypotheses in the induction steps a bit stronger, and such conditions
+can be added once we need them (for example in `h_add` it is only necessary to consider the sum of
+a simple function with a multiple of a characteristic function and that the intersection
+of their images is a subset of `{0}`. -/
+@[elab_as_elim]
+lemma Measurable.ennreal_sigmaFinite_induction [SigmaFinite μ] {P : (α → ℝ≥0∞) → Prop}
+    (h_ind : ∀ (c : ℝ≥0∞) ⦃s⦄, MeasurableSet s → μ s < ∞ → P (Set.indicator s fun _ => c))
```
Comment (RemyDegenne) [inline Mathlib/MeasureTheory/Function/SimpleFunc.lean:1235]:
```suggestion
    (h_ind : ∀ (c : ℝ≥0∞) ⦃s⦄, MeasurableSet s → μ s < ∞ → P (Set.indicator s fun _ ↦ c))
```

**Review by RemyDegenne — APPROVED** [review]
It looks good, let's merge this.
bors d+

**Comment by YaelDillies** [comment]
bors merge

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**